### ISSUE TITLE
[Baseline Reproduction] How to reproduce OLMo-3-7B Think/Instruct SFT -> solved

### ISSUE THREAD

**Issue by ferreirafabio** [allenai/open-instruct#1352]
Hey,
we attempted to reproduce the `OLMo-3-7B-Instruct-SFT` and `OLMo-3-7B-Think-SFT`checkpoints by fine-tuning the base model using the `olmo-sft.py` script from OLMo-core. We then evaluated our reproduced checkpoints against the official baselines using [OpenJury](https://github.com/OpenEuroLLM/OpenJury) (LLM-as-judge with `Qwen3-30B-A3B`).

Unfortunately, our reproduced models consistently underperform compared to the official baselines, and we're wondering if we missed any hyperparameters or training logic?

## Training Setup

  - **Base model**: OLMo-3-1025-7B (base checkpoint)
  - **Training script**: `olmo-sft.py` from OLMo-core
  - **Hyperparameters**: From Table 47 in the OLMo-3 paper (Section A.6.1)
    - Instruct: lr=8e-5 (trained for 2 epochs with batch size of 1M tokens, 32k max sequence length)
    - Think: lr=5e-5
  - **Data**: Dolci-Instruct-SFT and Dolci-Think-SFT datasets from HuggingFace
  - Hardware: trained on 8xH200
  

## Evaluation Results (OpenJury, winrate vs official baseline)

  | Model | Benchmark | Baseline | Ours | Δ |
  |-------|-----------|----------|------|---|
  | Instruct | alpaca-eval | 0.500 | 0.445 | -0.055 |
  | Instruct | arena-hard | 0.500 | 0.443 | -0.057 |
  | Think | alpaca-eval | 0.500 | 0.424 | -0.076 |
  | Think | arena-hard | 0.500 | 0.524 | +0.024 |

## Questions
1. Are there any hyperparameters used in the official training that are not reported in Table 47, e.g. which chat template did you use for think and instruct phases? (e.g. `olmo` (has function-calling system prompt) or `olmo_thinker_no_think_7b`?) For instance, the Instruct baseline has NO chat_template in its tokenizer config, so we don't know exactly what template was used for training

2. Is there any training logic (e.g., data preprocessing, chat template handling, tokenizer configuration) that is not present in `olmo-sft.py` but was used for the official checkpoints?

Thanks for your help.

Best,
Fabio

**Comment by BrownianNotion** [comment]
We also found a few points of [discrepancy in the evals](https://github.com/allenai/open-instruct/pull/1318#issuecomment-3778374147) when trying to reproduce the OLMo-2 1B SFT checkpoint. I've provided our reproduction versus the results reported by the paper below.

|         | Paper | Reproduction |
|---------|-------|-------|
| GSM8K   | 52.1  | 46.2  |
| IFEval  | 50.5  | 46.4  |

These results roughly match the open issue here https://github.com/allenai/open-instruct/issues/751. Also, cross-referencing https://github.com/allenai/olmes/issues/32#issuecomment-3787911869, https://github.com/allenai/open-instruct/issues/688 (issue closed as author did not reply).

The commands we used were
**Training Command:** 
This is simply the command given in `docs/olmo2.md` with the minor [path/arg fixes](https://github.com/allenai/open-instruct/pull/1318#issuecomment-3778374147) (eg. point at 1B model tokenizer, specify chat template).
```
accelerate launch \
    --mixed_precision bf16 \
    --num_processes 8 \
    --use_deepspeed \
    --deepspeed_config_file configs/ds_configs/stage3_no_offloading_accelerate.conf \
    --deepspeed_multinode_launcher standard \
    open_instruct/finetune.py \
    --exp_name olmo2_1b_v2_sft_lr3e-5_seed1  \
    --model_name_or_path allenai/OLMo-2-0425-1B \
    --model_revision main \
    --tokenizer_name allenai/OLMo-2-0425-1B \
    --tokenizer_revision main \
    --use_slow_tokenizer False \
    --add_bos \
    --chat_template_name tulu \
    --dataset_mixer_list allenai/tulu-3-sft-olmo-2-mixture-0225 1.0 \
    --use_flash_attn \
    --max_seq_length 4096 \
    --per_device_train_batch_size 2 \
    --gradient_accumulation_steps 8 \
    --learning_rate 3e-5 \
    --lr_scheduler_type linear \
    --warmup_ratio 0.03 \
    --weight_decay 0.0 \
    --num_train_epochs 2 \
    --report_to wandb \
    --with_tracking \
    --logging_steps 1 \
    --seed 1
```
The loss curves we obtained look quite similar to the original trajectory, except on a different scale

Our reproduction attempt

<img width="595" height="298" alt="Image" src="https://github.com/user-attachments/assets/20cbcb8b-43a2-46ef-9815-747ee93f005f" />

Original https://wandb.ai/ai2-llm/open_instruct_public/runs/mihc7mab

<img width="1652" height="626" alt="Image" src="https://github.com/user-attachments/assets/3c2c4708-7797-43d9-9e73-1b3935e200b4" />

**Eval Command:**
I did not use vLLM as it is currently broken on OLMES due to a version conflict https://github.com/allenai/olmes/issues/32.
```
olmes \
    --model Zettafleet/olmo2_1b_sft \
    --revision olmo2_1b_v2_sft_lr3e-5_seed1__1__1767915693 \
    --model-args '{"chat_model": true, "max_length": 4096, "metadata": {}}' \
    --task gsm8k::tulu \
    --batch-size 8 \
    --output-dir ours_gsm8k

olmes \
    --model Zettafleet/olmo2_1b_sft \
    --revision olmo2_1b_v2_sft_lr3e-5_seed1__1__1767915693 \
    --model-args '{"chat_model": true, "max_length": 4096, "metadata": {}}' \
    --task ifeval::tulu \
    --batch-size 4 \
    --output-dir ours_ifeval
```

Any insight or advice would be appreciated!

**Comment by BrownianNotion** [comment]
This is probably a red herring, but on the hugging face [model card](https://huggingface.co/allenai/OLMo-2-0425-1B-Instruct), it says that there is no bos token in the chat template for the 1B model. 
<img width="755" height="613" alt="Image" src="https://github.com/user-attachments/assets/a63881b5-40a8-415d-bf63-27a06f501a1e" />

However, when you load the 1B tokenizer for the SFT checkpoint, the bos_token is there
```
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("allenai/OLMo-2-0425-1B-SFT")
print("Chat template:", tokenizer.chat_template)

Chat template: {{ bos_token }}{% for message in messages %}{% if message['role'] == 'system' %}{{ '<|system|>
' + message['content'] + '
' }}{% elif message['role'] == 'user' %}{{ '<|user|>
' + message['content'] + '
' }}{% elif message['role'] == 'assistant' %}{% if not loop.last %}{{ '<|assistant|>
'  + message['content'] + eos_token + '
' }}{% else %}{{ '<|assistant|>
'  + message['content'] + eos_token }}{% endif %}{% endif %}{% if loop.last and add_generation_prompt %}{{ '<|assistant|>
' }}{% endif %}{% endfor %}
```

**Comment by ferreirafabio** [comment]
Hey, quick update. We managed to close the reproduction gap for both models. The issue was **the chat template**: our v1 runs used the base model's tokenizer as-is for the think stage, but the official checkpoints were trained with specific chat templates applied during preprocessing. Matching the template from the official HF checkpoints resolved the gaps.

| Model | v1 Avg Δ | v2 Avg Δ |
|---|---|---|
| Think SFT | -0.036 | **+0.009** |
| Instruct SFT | -0.056 | **+0.008** |

So the training script and hyperparameters from Table 47 are correct. 

Full evaluation report (winrate tables, rubric breakdowns, training curves): [eval_results.md](https://github.com/ferreirafabio/open-instruct/blob/main/oellm/evaluations/eval_results.md)

<img width="2225" height="1789" alt="Image" src="https://github.com/user-attachments/assets/3ab0bb1b-6078-4ff7-9030-929034cce731" />

**Comment by BrownianNotion** [comment]
@ferreirafabio Thanks for sharing! For OLMo-2 1B however, I think the chat template is already set correctly (I manually specified the `chat_template_name` flag with value tulu), and the template it maps to in `open_instruct/dataset_transformation.py` looks identical to the chat template from the SFT model on hugging face. So I don't think was the issue for OLMo-2 :(

**Comment by hamishivi** [comment]
Closing since this got solved, sorry for not looking deeply at it!
